File: python/euler/problems/p0004.py

# project euler: problem 4

from time import perf_counter

def is_palindrome(num):
    ns = str(num)

    return ns == ns[::-1]

def compute(limit):
    answer = []
    width = 10 ** 4
    for upper in range(limit, 0, -width):
        lower = upper - width
        for x in range(999, 0, -1):
            if x * x < lower:
                break
            for y in range(min(upper // x, x), 0, -1):
                tmp = x * y
                if tmp < lower:
                    break
                if is_palindrome(tmp) == True:
                    answer.append(tmp)

        if len(answer) != 0:
            return max(answer)

def solve():
    start = perf_counter()
    result = compute(10 ** 6)
    elapsed_time = perf_counter() - start

    return (result, "{:f}".format(elapsed_time))

File: clojure/src/project_euler/sol/p0004.clj
(ns project-euler.sol.p0004
  (:require [project-euler.lib.math :as my-math]))

;;; The search area is divided into blocks, and search blocks in descending order
;;; until the answer is found.

(defn solve
  ([]
   (solve 3))
  ([n-digits]
   {:pre [(pos? n-digits)]}
   (let [n-upper (dec (my-math/pow 10 n-digits))
         n-lower (if (> n-digits 1) (my-math/pow 10 (dec n-digits)) 0)
         blk-width (my-math/pow 10 (- (* n-digits 2) 2))]
     (loop [blk-lower-lst (range (* (quot (* n-upper n-upper) blk-width) blk-width) (dec (* n-lower n-lower)) (- blk-width))]
       (if (seq blk-lower-lst)
         (let [blk-lower (first blk-lower-lst)
               blk-upper (dec (+ blk-lower blk-width))
               prods (for [x (range n-upper (dec n-lower) -1)
                           :while (>= (* x x) blk-lower)
                           y (range (if (> x 0) (min (quot blk-upper x) x) x) (dec n-lower) -1)
                           :let [prod (* x y)]
                           :while (>= prod blk-lower)
                           :when (my-math/palindrome? prod)]
                       prod)]
           (if (seq prods)
             (apply max prods)
             (recur (next blk-lower-lst))))
         (throw (ex-info "No answer is found" {:n-digits n-digits})))))))
